text/0000-negative-bounds.md

```diff
@@ -0,0 +1,673 @@
+
+* **Negative trait bound** — `where T: !Int, U: !Eq<T>`
+* **Inequality bound** — `where T != u8`
+* **Negative outlives bound** — `where T: !'a, 'b: !'a`
+* **Negative projection bound** —  `where <T as Iterator>::Item != u8`
+
+> Note: `'b: !'a` is not the same as `'a: 'b`. In `'b: !'a`, the two lifetimes could merely be representing two irrelevant regions.
+
+There are three possible ways combine negation with higher-ranked modifier:
+
+* **Negated higher-ranked bound** — `where T: !for<'a> F<'a>`
+* **Higher-ranked negated bound** — `where T: for<'a> !F<'a>`
+* **Negated higher-ranked negated bound** — `where T: !for<'a> !F<'a>`
+
+In this RFC we will investigate all three cases, but as HRTB is already quite rare, we may choose to only support one of these forms.
+
+There is also a shorthand notation for projection bound:
+
+* `where T: Iterator<Item=u8>`, equivalent to
+* `where T: Iterator, T::Item = u8`
+
+The shorthand notation *cannot* be negated in the outermost position, because it would need a union in the desugared form.
+
+```rust
+// the short-hand notation of output constraint cannot be negated, since its
+// expansion cannot be written in the form `A + B + !C + !D + ...`.
+impl<T> Trait for T where T: !Iterator<Item=u8> { ... }
+// error: expected `,` or `>`, found `=`
+
+// The above can be simulated using two impls if desired:
+impl<T> Trait for T where T: !Iterator { ... }
+impl<T> Trait for T where T: Iterator, T::Item != u8 { ... }
+```
+
+We could allow negation in the inner position however:
+
+* `where T: Iterator<Item != u8>`, equivalent to
+* `where T: Iterator, T::Item != u8`
+
+Negation cannot be combined with relaxation. The following should not parse:
+
+```rust
+// negative relaxed bound reduces to an impossible bound, so it is pointless.
+struct NegativeRelaxedBound<T: !?Sized>;
+// error: expected identifier, lifetime or `for`, found `?`
+
+// relaxed negative bound and relaxed positive bound are just equivalent.
+struct RelaxedNegativeBound<T: ?!Sized>;
+// error: expected identifier, found `!`
+```
+
+## Evaluating disjoint collections
+
+### Trait bounds
+
+We could consider trait bounds to be sets, and concrete types satisfying the bound be elements in the set. Borrowing terminology from set theory, we call collections of bounds which have no common types be a *disjoint collection of bounds*. Bounds which are not disjoint are *overlapping*.
+
+In Rust, all traits are open for extension. Any traits with no superbounds are able to be implemented by the same type, even if the original developer may not think they should be used together. Therefore, all traits with no superbounds should be considered overlapping.
+
+The only way to create disjoint collection of trait bounds is by negative bounds. It is guaranteed that `!B` and `B` share no common types. In a collection `T: A + B + C + …`, as long as two of them are disjoint, the whole collection is also disjoint.
```

Comment:
I understand what you mean in the second sentence here, but the language could be clarified.